File: java/base/src/main/java/vsdk/toolkit/io/context/TransformScopeContext.java
package vsdk.toolkit.io.context;

import java.util.Arrays;

public class TransformScopeContext {
    public TransformStackContext transformContext;
    public String[] argumentList;
    public int argumentCount;
    public String iterateArgument;

    public TransformScopeContext() {
        transformContext = null;
        argumentList = null;
        argumentCount = 0;
        iterateArgument = "-i";
    }

    public void destroy() {
        clearArguments();
    }

    public void clearArguments() {
        argumentList = null;
        argumentCount = 0;
    }

    public int argumentCountFor(TransformStackContext context) {
        return context == null ? 0 : context.xac;
    }

    public int argumentStartIndexFor(TransformStackContext context) {
        return argumentCount - argumentCountFor(context);
    }

    public String[] argumentVectorFor(TransformStackContext context) {
        if (argumentList == null) {
            return null;
        }
        int startIndex = argumentStartIndexFor(context);
        if (startIndex < 0 || startIndex >= argumentList.length) {
            return null;
        }
        return Arrays.copyOfRange(argumentList, startIndex, argumentList.length);
    }

    public boolean compactTo(TransformStackContext context) {
        final int contextArgumentCount = argumentCountFor(context);
        String[] newArgumentList = null;

        if (contextArgumentCount > 0) {
            newArgumentList = new String[contextArgumentCount + 1];

            final int sourceStartIndex = argumentCount - contextArgumentCount;
            for (int i = 0; i < contextArgumentCount; i++) {
                newArgumentList[i] = argumentList[sourceStartIndex + i];
            }
            newArgumentList[contextArgumentCount] = null;
        }

        argumentList = newArgumentList;
        argumentCount = contextArgumentCount;
        return true;
    }

    public void freeTransformContext(TransformStackContext context) {
        if (context == null) {
            return;
        }
        if (context.ownedArgumentCopies != null) {
            for (int i = 0; i < context.ownedArgumentCount; i++) {
                context.ownedArgumentCopies[i] = null;
            }
            context.ownedArgumentCopies = null;
        }
        if (context.transformationArray != null) {
            context.transformationArray = null;
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/io/context/TransformScopeContext.ts
import { TransformStackContext } from "./TransformStackContext";

export class TransformScopeContext {
  public transformContext: TransformStackContext | null;
  public argumentList: Array<string | null> | null;
  public argumentCount: number;
  public iterateArgument: string;

  public constructor() {
    this.transformContext = null;
    this.argumentList = null;
    this.argumentCount = 0;
    this.iterateArgument = "-i";
  }

  public destroy(): void {
    this.clearArguments();
  }

  public clearArguments(): void {
    this.argumentList = null;
    this.argumentCount = 0;
  }

  public argumentCountFor(context: TransformStackContext | null): number {
    return context === null ? 0 : context.xac;
  }

  public argumentStartIndexFor(context: TransformStackContext | null): number {
    return this.argumentCount - this.argumentCountFor(context);
  }

  public argumentVectorFor(context: TransformStackContext | null): Array<string | null> | null {
    if (this.argumentList === null) {
      return null;
    }
    const startIndex = this.argumentStartIndexFor(context);
    if (startIndex < 0 || startIndex >= this.argumentList.length) {
      return null;
    }

    const copy = new Array<string | null>(this.argumentList.length - startIndex);
    for (let i = 0; i < copy.length; i++) {
      copy[i] = this.argumentList[startIndex + i] ?? null;
    }
    return copy;
  }

  public compactTo(context: TransformStackContext | null): boolean {
    const contextArgumentCount = this.argumentCountFor(context);
    let newArgumentList: Array<string | null> | null = null;

    if (contextArgumentCount > 0) {
      newArgumentList = new Array<string | null>(contextArgumentCount + 1).fill(null);
      const sourceStartIndex = this.argumentCount - contextArgumentCount;
      for (let i = 0; i < contextArgumentCount; i++) {
        newArgumentList[i] = (this.argumentList as Array<string | null>)[sourceStartIndex + i] ?? null;
      }
      newArgumentList[contextArgumentCount] = null;
    }

    this.argumentList = newArgumentList;
    this.argumentCount = contextArgumentCount;
    return true;
  }

  public freeTransformContext(context: TransformStackContext | null): void {
    if (context === null) {
      return;
    }
    if (context.ownedArgumentCopies !== null) {
      for (let i = 0; i < context.ownedArgumentCount; i++) {
        context.ownedArgumentCopies[i] = null;
      }
      context.ownedArgumentCopies = null;
    }
    if (context.transformationArray !== null) {
      context.transformationArray = null;
    }
  }
}
